modules/developers-guide/pages/tutorials/reproducible-builds.adoc

```diff
@@ -0,0 +1,341 @@
+In particular, if you are a canister user as opposed to the owner, you likely want to verify that the canister is running some intended code before giving it control over your ICP tokens or confidential data.
+Verifying this requires answering two questions:
+
+. Which WebAssembly (WASM) code is being executed for a canister? 
+. The canisters are normally written in a higher-level language, such as Motoko or Rust, and not directly in WASM.
+   The question is then: is the WASM that's running really the result of compiling the purported source code? 
+
+The rest of the document answers these questions.
+For the first question, we will see how the {IC} provides information about canister code.
+To be able to answer the second question, canister authors should ensure trusted and 
+https://reproducible-builds.org/docs/definition/[reproducible builds] of the WASM code from the source.
+Such builds allows anyone to follow the same steps as the canister authors and yield the exact same WASM code,
+which can then be compared to the code executing on the {IC}.
+
+The impatient reader who is familiar with the topic of reproducible builds can skip straight to the <<repro-build-summary,summary>>.
+
+== Finding out which code the {IC} is executing
+
+{IC} does not allow you to access the WASM code of an arbitrary canister.
+This is a design decision, as developers might want to keep some code private.
+However, the {IC} does allow you to access the _cryptographic hash_ of the WASM code of a canister.
+More concretely, it provides you the SHA-256 hash of the code.
+
+To obtain this hash, you must first note the ID of the {IC} canister whose code you want to check.
+For example, assume we're interested in the code of the Internet Identity canister, whose ID is `rdmx6-jaaaa-aaaaa-aaadq-cai`.
+Then, the easiest way to access this service is using the `dfx` tool from the terminal.
+Open your terminal, and run:
+
+----
+$ dfx canister --no-wallet --network ic info rdmx6-jaaaa-aaaaa-aaadq-cai
+Controller: r7inp-6aaaa-aaaaa-aaabq-cai
+Module hash: 0xda8d1bdd93fbf1edddeb2a423fa3c528c2e4877d39beed2277a12e60227e44d4
+----
+
+NOTE: If you are running an older version of `dfx`, you will need to run this command from a directory that contains a valid `dfx.json` file.
+If you don't have such a directory, you can create it using `dfx new`.
+
+Here, the {IC} tells us that the hash of the code of the `rdmx6-jaaaa-aaaaa-aaadq-cai` canister (which happens to be the internet identity canister) is `0xda8d1bdd93fbf1edddeb2a423fa3c528c2e4877d39beed2277a12e60227e44d4`.
+
+CAUTION: The check above provides you the _current_ hash of the canister, but the _controllers_ of an {IC} canister may change the code at any time (e.g., to upgrade a canister). Currently, there is only way to ensure that the canister hasn't changed its code since you last checked it: the canister's controllers must be empty, preventing the code from upgrading.
```

Comment:
```suggestion
CAUTION: The check above provides you the _current_ hash of the canister, but the _controllers_ of an {IC} canister may change the code at any time (e.g., to upgrade a canister). However, if the list of controllers is empty, you know that the canister is immutable since nobody has the power to change the code.
```